core/arch/arm/plat-k3/drivers/mailbox.c

```
@@ -0,0 +1,192 @@
+
+/*
+ * TI Mailbox MMIO and SRAM base addresses
+ * These are the base addresses for the mailbox MMIO and SRAM regions.
+ * They are used to map the physical addresses to virtual addresses.
+ */
+
+static void *mailbox_rx_base;
+static void *mailbox_tx_base;
+static void *mailbox_tx_sram_va;
+static void *mailbox_rx_sram_va;
+
+/*
+ * mailbox_init() - Initialize the mailbox MMIO and SRAM regions
+ */
+
+void mailbox_init(void)
+{
+	mailbox_rx_base = (void *)core_mmu_get_va(TI_MAILBOX_RX_BASE, MEM_AREA_IO_SEC, 0x1000);
+	mailbox_tx_base = (void *)core_mmu_get_va(TI_MAILBOX_TX_BASE, MEM_AREA_IO_SEC, 0x1000);
+	mailbox_tx_sram_va = (void *)core_mmu_get_va(MAILBOX_TX_START_REGION, MEM_AREA_IO_SEC, 0x1000);
+	mailbox_rx_sram_va = (void *)core_mmu_get_va(MAILBOX_RX_START_REGION, MEM_AREA_IO_SEC, 0x1000);
+
+	if (!mailbox_rx_base || !mailbox_tx_base || !mailbox_tx_sram_va || !mailbox_rx_sram_va) {
+		EMSG("Failed to map mailbox MMIO or SRAM regions: rx=%p tx=%p tx_sram=%p rx_sram=%p",
+		     mailbox_rx_base, mailbox_tx_base, mailbox_tx_sram_va, mailbox_rx_sram_va);
+		return;
+	}
+
+	IMSG("Mailbox MMIO and SRAM regions mapped: rx=%p tx=%p tx_sram=%p rx_sram=%p",
+	      mailbox_rx_base, mailbox_tx_base, mailbox_tx_sram_va, mailbox_rx_sram_va);
+}
+
+/*
+ * Function to poll for mailbox rx messages
+ */
+
+static int8_t ti_mailbox_poll_rx_status(void)
+{
+	uint32_t num_messages_pending = 0U;
+	uint32_t retry_count = 100U;
+
+	while (num_messages_pending == 0U) {
+		num_messages_pending = io_read32((vaddr_t)mailbox_rx_base + TI_MAILBOX_MSG_STATUS);
+
+		if (retry_count-- == 0U) {
+			EMSG("Mailbox RX status polling timed out");
+			return -1;
+		}
+		udelay(10000);
+	}
+
+	return 0;
+}
+
+/*
+ * ti_sci_transport_send() - Send data over a TISCI transport
+ */
+
+TEE_Result ti_sci_transport_send(const struct ti_sci_msg *msg)
+{
+	uint32_t num_bytes;
+	void *dst_ptr = mailbox_tx_sram_va;
+	paddr_t pa;
+
+	if (!msg)
+		return -1;
```

Return error codes (TEE_SUCCESS or TEE_ERROR_*) instead of 0 or -1 in all places functions returning a TEE_Result